frontend/src/components/UserProfile/UserName.tsx

```diff
@@ -0,0 +1,180 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import {
+  Alert,
+  Control,
+  Field,
+  Label,
+  Root,
+  Button,
+} from "@vector-im/compound-web";
+import { useAtomValue, useAtom, useSetAtom } from "jotai";
+import { atomFamily } from "jotai/utils";
+import { atomWithQuery, atomWithMutation } from "jotai-urql";
+import {
+  useState,
+  useEffect,
+  ChangeEventHandler,
+  FormEventHandler,
+} from "react";
+
+import { graphql } from "../../gql";
+import LoadingSpinner from "../LoadingSpinner/LoadingSpinner";
+
+import styles from "./UserName.module.css";
+
+const QUERY = graphql(/* GraphQL */ `
+  query UserGreeting($userId: ID!) {
+    user(id: $userId) {
+      id
+      username
+      matrix {
+        mxid
+        displayName
+      }
+    }
+  }
+`);
+
+const userGreetingFamily = atomFamily((userId: string) => {
+  const userGreeting = atomWithQuery({
+    query: QUERY,
+    getVariables: () => ({ userId }),
+  });
+
+  return userGreeting;
+});
+
+const SET_DISPLAYNAME_MUTATION = graphql(/* GraphQL */ `
+  mutation SetDisplayName($userId: ID!, $displayName: String!) {
+    setDisplayName(input: { userId: $userId, displayName: $displayName }) {
+      status
+      user {
+        id
+        matrix {
+          displayName
+        }
+      }
+    }
+  }
+`);
+
+const setDisplayNameAtom = atomWithMutation(SET_DISPLAYNAME_MUTATION);
+
+const getErrorMessage = (result: {
+  error?: unknown;
+  data?: { setDisplayName: { status: string } };
+}): string | undefined => {
+  if (result.error) {
+    return "Failed to save display name. Please try again.";
+  }
+  if (result.data?.setDisplayName.status === "INVALID") {
+    return "Failed to save invalid display name.";
+  }
+};
+
+const UserName: React.FC<{ userId: string }> = ({ userId }) => {
+  const result = useAtomValue(userGreetingFamily(userId));
+
+  const [setDisplayNameResult, setDisplayName] = useAtom(setDisplayNameAtom);
+  const [inProgress, setInProgress] = useState(false);
+
+  const user = result.data?.user;
+  const displayName = user?.matrix.displayName || "";
+
+  const [editingDisplayName, setEditingDisplayName] = useState(displayName);
+
+  const userGreeting = useSetAtom(userGreetingFamily(userId));
+
+  useEffect(() => {
+    setEditingDisplayName(displayName);
+  }, [displayName]);
+
+  const onDisplayNameChange: ChangeEventHandler<HTMLInputElement> = (
+    event,
+  ): void => {
+    setEditingDisplayName(event.target.value);
+  };
```

Comment:
I'm not a fan of state-controlled input like that. I prefer to leave it uncontrolled, and grab the value in the form on submit using a ref, and reset the field with `HTMLFormElement#reset()`

There is an example of that here: https://github.com/matrix-org/matrix-authentication-service/blob/fd22d6e26ec31f871810ccec86c58e9fa8b0dc0b/frontend/src/components/VerifyEmail/VerifyEmail.tsx#L120-L137